### ISSUE TITLE
Increased memory usage from 10Mo to 200Mo

### ISSUE THREAD

**Issue by geo42** [mulesoft/osprey#139]
Last week I downloaded my node modules from scratch and my application used 200 extra Mo of RAM.

So i made in simple node app with a single line : require('osprey'); (which is in version 0.3.2)

Usually on my device, osprey uses around 10MB or RAM, and now it's around 200MB

I checked the module dependencies, and i think the problem comes from 
raml-path-match@2.1.0 and
raml-validate@1.1.0

When i delete those two directories (rm -rf) and install raml-path-match@2.0.0 and raml-validate@1.0.6, memory usage is back to normal.

**Comment by jstoiko** [comment]
Thanks for reporting. 

I'm assuming your RAML is 0.8 since you are using `Osprey@0.3.2`. I would be interested to know what happens when you use `Osprey@0.4.0` (latest release) with that same RAML.

Then I would also be interested to know what happens when you convert your RAML to 1.0 and use `Osprey@0.4.0`.

I hope I'm not asking too much. This would be super useful!

**Comment by geo42** [comment]
I'm not using any RAML, I use a test file and only require osprey. I'm not doing anything with osprey.

**Comment by jstoiko** [comment]
what do you pass to Osprey?

**Comment by geo42** [comment]
This is the minimal code I run : 

`var osprey = require('osprey');

var cb = function () {
 console.log('fin');
}

setTimeout(cb, 10000);`

I pass nothing to osprey, the only thing I do is require it

Same behavior unsing osprey 0.4.0
My node version is 6.9.3

**Comment by geo42** [comment]
Did you get this behaviour as well ?

**Comment by jstoiko** [comment]
I couldn't reproduce. 

```sh
$ node --expose-gc
> process.memoryUsage()
{ rss: 22671360, heapTotal: 10295296, heapUsed: 5509832 }
> var osprey = require('osprey')
undefined
> process.memoryUsage()
{ rss: 91172864, heapTotal: 51883520, heapUsed: 41322072 }

```

What are you using to measure memory?

**Comment by geo42** [comment]
Using the same test as yours : 

node --expose-gc
> process.memoryUsage()
{ rss: 27435008, heapTotal: 11571200, heapUsed: 5737312 }

> var osprey = require('osprey')
{ rss: 277581824, heapTotal: 104144896, heapUsed: 71367736 }

I'm usually using a lts version of Node on a Debian 8.0 using virtualbox, and aws
The previous result is displayed on a 6.10.0

With a 7.7.2 :

node --expose-gc
> process.memoryUsage()
{ rss: 28975104,
  heapTotal: 7258112,
  heapUsed: 4560896,
  external: 32146 }
> var osprey = require('osprey')
undefined
> process.memoryUsage()
{ rss: 213647360,
  heapTotal: 105586688,
  heapUsed: 73712696,
  external: 305971 }

Before the test I deleted my node_mudules and performed a npm cache clean, before npm i osprey again (0.4.0)

**Comment by geo42** [comment]
Does it help ? What extra information should I provide ?

**Comment by jstoiko** [comment]
I still can't reproduce the 200MB spike even with the same node version. I get a 100MB spike and then it starts going down after about 30sec, 90MB, after 2-3 min, it's down do 60MB, etc.

Does it ever go down on your environment?

**Comment by geo42** [comment]
I was surprised that you could not reproduce, so I've been investigating a bit further using several OS on the latest node version : 

windows 7
node 7.7.3

C:\Users\IEUser\Desktop\node>node
> process.memoryUsage()
{ rss: 16502784,
heapTotal: 4886528,
heapUsed: 2801180,
external: 9311 }
> var osprey = require("osprey")
undefined
> process.memoryUsage()
{ rss: 79564800,
heapTotal: 63905792,
heapUsed: 49870048,
external: 543810 }

After a while...
> process.memoryUsage()
{ rss: 63475712,
heapTotal: 48574464,
heapUsed: 26733972,
external: 543818 }

Which seems Ok

OSX 10.12.3
node 7.7.3

> process.memoryUsage()
{ rss: 23896064,
  heapTotal: 7258112,
  heapUsed: 4152288,
  external: 9894 }
> process.memoryUsage()
{ rss: 138752000,
  heapTotal: 102965248,
  heapUsed: 74261872,
  external: 966940 }

After a while...
> process.memoryUsage()
{ rss: 53964800,
  heapTotal: 102965248,
  heapUsed: 74413336,
  external: 966952 }

It seems to need a lot of memory at start (138MB) and release it after a few minutes

Linux Debian Jessie
node 7.7.3

> process.memoryUsage()
{ rss: 29454336,
  heapTotal: 7258112,
  heapUsed: 5097048,
  external: 32146 }
> var osprey = require('osprey')
undefined
> process.memoryUsage()
{ rss: 204738560,
  heapTotal: 98127872,
  heapUsed: 74401424,
  external: 958744 }

After a while...
> process.memoryUsage()
{ rss: 149123072,
  heapTotal: 41504768,
  heapUsed: 32481744,
  external: 25685 }

It uses a lot more memory (204MB), after a while it releases around 50MB, but a lot remains (149MB)


In conclusion, it seems to me that some memory is retained then released in Win32 an OSX but not on the linux version.
Furthermore, OSX seems to retain a lot of memory at start, as compared to Win32.
Linux uses the most memory and fails to release it.

**Comment by jstoiko** [comment]
Any chance you could redo the same test after uninstalling the `raml-1-parser` and `raml-typesystem`? These are the only two dependencies that have been added to `Osprey`. It should let you do a `require()` without those dependencies.

**Comment by geo42** [comment]
Hello,

On the linux Machine, I first removed `raml-1-parser`, the mory kept going up 200MB.
After removing `raml-typesystem`, I could not require osprey anymore.

Error: Cannot find module 'raml-typesystem'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/developer/tmp/node_modules/raml-validate/raml-validate.js:2:22)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)

As I told you on my first post here, I have been able to get back to the previous behaviour (60MB) by downgrading to the 2 modules / versions : raml-path-match@2.0.0 and raml-validate@1.0.6

Can you reproduce this behaviour using a Linux device of your own ?

**Comment by geo42** [comment]
I saw that the new version osprey-router@0.5.0 was available, so I tried it with osprey@0.4.0
It keeps using the same amount of memory on linux.

**Comment by jstoiko** [comment]
I'm not sure what's causing this.

If I run the same test by requiring `raml-path-match` and `raml-validate` individually, the memory footprint is fine.

Can you try again after running `npm dedupe` and see if that helps. Also you can try upgrading the parser - I believe the latest version is `1.1.9` - to see if that helps too.

**Comment by geo42** [comment]
I'm was currently using the latest version : raml-1-parser@1.1.19.

When I load raml-validate by itself, memory is up to 130 MB

> process.memoryUsage()
{ rss: 28753920,
  heapTotal: 7258112,
  heapUsed: 4472368,
  external: 32124 }
> var rv = require('raml-validate')
undefined
> process.memoryUsage()
{ rss: 129048576,
  heapTotal: 28839936,
  heapUsed: 20844120,
  external: 183168 }

When I load raml-path-match : 130 MB

> process.memoryUsage()
{ rss: 28766208,
  heapTotal: 7258112,
  heapUsed: 4444328,
  external: 32136 }
> var rpm = require('raml-path-match')

undefined
> process.memoryUsage()
{ rss: 128778240,
  heapTotal: 24645632,
  heapUsed: 20926384,
  external: 183168 }

When I load both : 130MB

> process.memoryUsage()
{ rss: 28610560,
  heapTotal: 7258112,
  heapUsed: 4433712,
  external: 32130 }
> var rpm = require('raml-path-match')
undefined
> process.memoryUsage()
{ rss: 127897600,
  heapTotal: 28315648,
  heapUsed: 21974120,
  external: 283745 }
> var rv = require('raml-validate')
undefined
> process.memoryUsage()
{ rss: 128167936,
  heapTotal: 28315648,
  heapUsed: 22348784,
  external: 283771 }

Both, the other way : 130MB

> process.memoryUsage()
{ rss: 28729344,
  heapTotal: 7258112,
  heapUsed: 4433712,
  external: 32130 }
> var rv = require('raml-validate')
undefined
> process.memoryUsage()
{ rss: 128819200,
  heapTotal: 28839936,
  heapUsed: 20809272,
  external: 183180 }
> var rpm = require('raml-path-match')
undefined
> process.memoryUsage()
{ rss: 129085440,
  heapTotal: 28839936,
  heapUsed: 21330592,
  external: 195923 }

When I run npm dedupe and require osprey, it still is up to 200MB then drops to 140MB after a while
Same behavior if i don't run npm dedupe.

What are your footprints on linux when you load these libraries ?

**Comment by jstoiko** [comment]
Closing this issue as there hasn't been any activity in a while. Feel free to re-open.
